utils/zerovec/src/varzerovec/ule.rs

```diff
@@ -0,0 +1,161 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::components::SliceComponents;
+use super::*;
+use core::marker::PhantomData;
+use core::mem;
+
+/// A VarULE version of VarZeroVec allowing for VarZeroVecs to be nested indefinitely
+///
+/// ```rust
+/// use zerovec::VarZeroVec;
+/// use zerovec::ZeroVec;
+/// use zerovec::varzerovec::VarZeroVecULE;
+/// use zerovec::ule::*;
+/// let strings_1: Vec<String> = vec!["foo".into(), "bar".into(), "baz".into()];
+/// let strings_2: Vec<String> = vec!["twelve".into(), "seventeen".into(), "forty two".into()];
+/// let strings_3: Vec<String> = vec!["我".into(), "喜歡".into(), "烏龍茶".into()];
+/// let strings_4: Vec<String> = vec!["w".into(), "ω".into(), "文".into(), "𑄃".into()];
+/// let strings_12 = vec![strings_1.clone(), strings_2.clone()];
+/// let strings_34 = vec![strings_3.clone(), strings_4.clone()];
+/// let all_strings = vec![strings_12, strings_34];
+///
+/// let vzv_1 = VarZeroVec::from(&*strings_1);
+/// let vzv_2 = VarZeroVec::from(&*strings_2);
+/// let vzv_3 = VarZeroVec::from(&*strings_3);
+/// let vzv_4 = VarZeroVec::from(&*strings_4);
+/// let vzv_12 = VarZeroVec::from(&[vzv_1, vzv_2] as &[_]);
+/// let vzv_34 = VarZeroVec::from(&[vzv_3, vzv_4] as &[_]);
+/// let vzv_all = VarZeroVec::from(&[vzv_12, vzv_34] as &[_]);
+///
+/// let reconstructed = vzv_all.iter()
+///        .map(|v: &VarZeroVecULE<_>| {
+///             v.iter().map(|x: &VarZeroVecULE<_>| x.as_varzerovec().to_vec()).collect::<Vec<_>>()
+///         }).collect::<Vec<_>>();
+/// assert_eq!(reconstructed, all_strings);
+///
+/// let bytes = vzv_all.get_encoded_slice();
+/// let vzv_from_bytes: VarZeroVec<VarZeroVec<VarZeroVec<String>>> = VarZeroVec::try_from_bytes(bytes).unwrap();
+/// assert_eq!(vzv_from_bytes, vzv_all);
+/// ```
+//
+// safety invariant: The slice MUST be one which parses to
+// a valid SliceComponents<T>
+#[repr(transparent)]
+pub struct VarZeroVecULE<T> {
+    marker: PhantomData<[T]>,
+    /// The original slice this was constructed from
+    entire_slice: [u8],
```

Comment:
Not quite because that's not growable. But yes I'd have loved to unify the two, I _could_ use some kind of trait that encapsulates `.get_components()` but it's not really worth it IMO because that pollutes the public API with a confusing generic type.